Handle errors on crontasks

cedric-anne opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Here is a simple solution to prevent cron tasks to be marked as running when an error occured. Recoverable errors will be catched and so the task execution will be considered as aborted.
Consequences are:
 - task will be reexecuted on next scheduled date,
 - scripts will not exit due to error, so following tasks on the loop will be executed.

A better solution would be to add the error trace in the `glpi_crontasklogs` table, but it would require to add a new field in DB, much new strings and some UI changes, so it cannot be done on a bugfixes branch.

I forced an error on mailgate, for the example.

Cron log:
```
2021-03-26 09:26:01 [@69509cfe24fe]
External #1: Launch mailgate
2021-03-26 09:26:02 [@69509cfe24fe]
External #1: Error during mailgate execution. Check in "/var/www/glpi/files/_log/php-errors.log" for more details.
2021-03-26 09:26:02 [@69509cfe24fe]
External #2: Launch slaticket
2021-03-26 09:26:03 [@69509cfe24fe]
External #3: Launch planningrecall
2021-03-26 09:26:03 [@69509cfe24fe]
External #4: Launch queuednotification
2021-03-26 09:26:03 [@69509cfe24fe]
External #5: Launch olaticket
```

PHP errors log:
```
[2021-03-26 09:26:02] glpiphplog.CRITICAL:   *** Uncaught Exception Error: Call to a member function next() on null in /var/www/glpi/inc/mailcollector.class.php at line 1738
  Backtrace :
  inc/crontask.class.php:848                         MailCollector::cronMailgate()
  front/cron.php:83                                  CronTask::launch()
```

Cron task log:
![image](https://user-images.githubusercontent.com/33253653/112611223-d7c2dd80-8e1d-11eb-818d-c7cc4fe4448b.png)

orthagh commented on inc/crontask.class.php:
```diff
@@ -836,17 +840,30 @@ static public function launch($mode, $max = 1, $name = '') {
                                                 __('Launch'), $crontask->fields['name']));
                      Toolbox::logInFile('cron', $msgcron);
-                     $retcode = call_user_func($fonction, $crontask);
+                     try {
```
can you take the opportunity to rename `fonction` to `function` ?

cedric-anne commented:
To be sure that this will not generate excessive logs, we have to block task execution after X successive failures.

In order to do this we could:
 - add a log with a new `error` state on `CrontaskLog`;
 - switch the `Crontask` to a new `errored` state after X successive failures;
 - create a notification that would be sent by default to the admin when this happens.

cedric-anne commented:
I added a mechanism to send a notification to administrator whe a task fails 5 times in the last 10 executions. This notification will not be sent more than once a day and uses the existing notification used by `CronTask::cronWatcher()`.
![image](https://user-images.githubusercontent.com/33253653/112796371-2e245c00-906a-11eb-9093-a4a97dba4fb6.png)

I also added a `CronTaskLog::STATE_ERROR`, to have make the log more clear:
![image](https://user-images.githubusercontent.com/33253653/112796479-514f0b80-906a-11eb-99c8-8d3c1f125379.png)


Finally, I choosed to not implement an error state on the `CronTask` class itself, has it has many impacts and should not be done on a bugfixes branch.
In the worst case, the crontask will fail every minute and generate a 1kB log, so about 1.5MB of logs for a whole day. It could be a problem only if administrator ignore the notification for a very long time and has not set any log rotation.
